// game/SecondScene.js
import 'game/style.css'; // webpack magic
import * as input from 'engine/input';
import { Point, SCALE_MODES, settings, Sprite, Rectangle } from 'pixi.js';

import * as engine from 'engine';
import * as display from 'engine/display';
import * as resources from 'engine/resources';
import Scene from 'engine/Scene';
import { Head, Food, FakeHead } from 'game/Snake';

export class SecondScene extends Scene {
  init() {
    this.backgroundColor = 0xEEEEEE;

    const ele = document.getElementById("Root");
    ele.style.width = display.width;
    ele.style.height = display.height;

    const center = new Point(display.width / 2, display.height / 2);
    const topLeft = new Point(display.width * 0.1, display.height * 0.1);

    this.character = new Head(center, this.stage);
    this.fake = new FakeHead(new Point(display.width * 0.5, display.height * 0.5), this.stage);
    this.fake.control.delta = Math.PI * 0.5;

    this.bound = new Rectangle(0, 0, display.width, display.height);

    this.enemy = [];
  }

  endGame() {
    const center = new Point(display.width / 2, display.height / 2);
    this.character.reset(center);
  }

  update(dt) {
    this.character.update(dt);
    this.fake.update(dt);

    if (!this.bound.contains(
      this.character.sprite.position.x,
      this.character.sprite.position.y)) {
      this.endGame();
    }

    if (!this.bound.contains(
      this.fake.sprite.position.x,
      this.fake.sprite.position.y)) {
      if (!this.fake.isLocked) {
        this.fake.auto.updateDest();
        this.fake.isLocked = true;
      }
    } else {
      this.fake.isLocked = false;
    }

    if (this.fake.checkHit(this.character.sprite)) {
      this.endGame();
    }

    //if(input.buttonReleased('A')){
    //  this.enemy.push(new FakeHead());
    //}
  }
}
