// game/Map.js
import { Point } from "pixi.js";
import { Enemy } from 'game/Soldier';

export class Map {
    constructor(width, height) {
        this.waveId = 0;
        this.size = new Point(width, height);
    }

    spawnEnemy(container) {
        const list = [];

        if (this.waveId >= MapConfig.length)
            return list;

        const wave = MapConfig[this.waveId++];
        const spawnPos = new Point;

        for (let i = 0; i < wave.length; i++) {
            const data = wave[i];

            spawnPos.set(this.size.x * data.x, this.size.y * data.y);
            const e = new Enemy(spawnPos, container);
            e.sprite.v.set(data.vX, data.vY);
            e.sprite.a.set(data.aX, data.aY);
            e.sprite.tint = color[i];
            list.push(e);
        }

        return list;
    }
}

const color = [0xFF00FF, 0xFFFF00, 0x00FFFF, 0x00FFFF, 0x00FFFF, 0x00FFFF];

const MapConfig =
    [
        [
            {
                x: 0.8, y: 0.8,
                aX: 0, aY: 0,
                vX: -50, vY: 0,
            },
        ],
    ];
